cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/preview/DefaultPreviewManager.java

```diff
@@ -16,55 +16,242 @@
+import java.util.HashSet;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.ParametersAreNonnullByDefault;
 
 /**
- * Default implementation of the {@link PreviewManager}.
+ * Class responsible for creating the injector for preview and starting it.
  */
 public class DefaultPreviewManager implements PreviewManager {
-  @Override
-  public ApplicationId start(NamespaceId namespaceId, AppRequest<?> request) throws Exception {
-    return null;
-  }
 
-  @Override
-  public PreviewStatus getStatus(ApplicationId preview) throws NotFoundException {
-    return null;
-  }
+  private static final Logger LOG = LoggerFactory.getLogger(DefaultPreviewManager.class);
+  private final CConfiguration cConf;
+  private final Configuration hConf;
+  private final DiscoveryService discoveryService;
+  private final DatasetFramework datasetFramework;
+  private final PreferencesStore preferencesStore;
+  private final SecureStore secureStore;
+  private final TransactionManager transactionManager;
+  private final ArtifactRepository artifactRepository;
+  private final ArtifactStore artifactStore;
+  private final AuthorizerInstantiator authorizerInstantiator;
+  private final StreamAdmin streamAdmin;
+  private final StreamCoordinatorClient streamCoordinatorClient;
+  private final PrivilegesManager privilegesManager;
+  private final AuthorizationEnforcer authorizationEnforcer;
+  private final Cache<ApplicationId, Injector> appInjectors;
 
-  @Override
-  public void stop(ApplicationId preview) throws Exception {
+  @Inject
+  DefaultPreviewManager(final CConfiguration cConf, Configuration hConf, DiscoveryService discoveryService,
+                        @Named(DataSetsModules.BASE_DATASET_FRAMEWORK) DatasetFramework datasetFramework,
+                        PreferencesStore preferencesStore, SecureStore secureStore,
+                        TransactionManager transactionManager, ArtifactRepository artifactRepository,
+                        ArtifactStore artifactStore, AuthorizerInstantiator authorizerInstantiator,
+                        StreamAdmin streamAdmin, StreamCoordinatorClient streamCoordinatorClient,
+                        PrivilegesManager privilegesManager, AuthorizationEnforcer authorizationEnforcer) {
+    this.cConf = cConf;
+    this.hConf = hConf;
+    this.datasetFramework = datasetFramework;
+    this.discoveryService = discoveryService;
+    this.preferencesStore = preferencesStore;
+    this.secureStore = secureStore;
+    this.transactionManager = transactionManager;
+    this.artifactRepository = artifactRepository;
+    this.artifactStore = artifactStore;
+    this.authorizerInstantiator = authorizerInstantiator;
+    this.streamAdmin = streamAdmin;
+    this.streamCoordinatorClient = streamCoordinatorClient;
+    this.privilegesManager = privilegesManager;
+    this.authorizationEnforcer = authorizationEnforcer;
+
+    // TODO make maximum size and expire after write configurable?
```

Comment:
It's probably ok. If you really worry about OOM issue, use weakKey so that when GC kicks in, the entry will be subject to GC (at the least the key and you have chance to get the value and stop the preview runner).